Back up edited `full` image sources when restoring the original image

mitogh opened the pull request:
## Summary

<!-- Please reference the issue this PR fixes. If this PR does not fix the entire issue, change this to Addresses #... instead. -->
Store the sources and update the `_wp_attachment_backup_sources` meta when the image is restored, only update the meta if the meta location has not been updated yet.

Fixes #295 

## Relevant technical choices

A new class was introduced when the constant `IMAGE_EDIT_OVERWITE` is set to true due this needs to happen before WordPress has been bootrstaped, this class can be used as time move forwards for any additional scenario when this constant is defined as `true` and allowing the rest of the other tests to behave in a scenario when the constant is not defined. `@runInIsolation` can't be used as part of the tests due we are using clousures and the clousures can't be serialized when running a test in isolaion.

The backup tries to find the edited image when the image is restored in order to store it in the same way as when the image is edited. 

<!-- Please describe your changes. -->

## Checklist

- [x] PR has either `[Focus]` or `Infrastructure` label.
- [x] PR has a `[Type]` label.
- [x] PR has a milestone or the `no milestone` label.

mitogh committed:
Create set of tests to replicate the reported bug

When an image is edited and restored, the full size sources
are not stored in the meta `_wp_attachment_backup_sources` with
this change we introduce the storage of those values when
the constant IMAGE_EDIT_OVERWRITE is not defined or it contains
a falsy value.

In order to follow the same patter WordPress uses to store the backup
sizes.

mitogh committed:
Backup the sources for additional targets

When restoring an image, make sure any edited source for the full
size image is stored in the meta used to backup the sources, in the same
way it happens for the sizes when the constant `IMAGE_EDIT_OVERWRITE` is
defined.

felixarntz commented on modules/images/webp-uploads/image-edit.php:
```diff
@@ -320,10 +320,39 @@ function webp_uploads_get_next_full_size_key_from_backup( $attachment_id ) {
+	}
+
+	if ( ! defined( 'IMAGE_EDIT_OVERWRITE' ) || ! IMAGE_EDIT_OVERWRITE ) {
+		$target = null;
```
Maybe let's call this variable `$suffix`? `$target` is more commonly used for the size target (e.g. "thumbnail"). WordPress core calls this `$suffix`, so I think it's better to align with that here for more clarity. While unrelated, I suggest we also adjust that in the existing code for `webp_uploads_backup_full_image_sources()`.

mitogh replied:
Sure, I can rename it, the main reason behind this was called `target` is because it's the target suffix we are going to use from the existing ones.

felixarntz commented on modules/images/webp-uploads/image-edit.php:
```diff
@@ -320,10 +320,39 @@ function webp_uploads_get_next_full_size_key_from_backup( $attachment_id ) {
  */
 function webp_uploads_restore_image( $attachment_id, $data ) {
-	$backup_sources = get_post_meta( $attachment_id, '_wp_attachment_backup_sources', true );
+	if ( empty( $data['sources'] ) ) {
+		return $data;
+	}
```
How can we expect `sources` to be set here? This hook is called from core's `wp_update_attachment_metadata()` call, and core doesn't set `sources` within the metadata. I think we need to look at the _previous_ attachment metadata here via `wp_get_attachment_metadata( $attachment_id )`, similar to how we do it already in our `webp_uploads_backup_sources()` implementation.

mitogh replied:
Good question, the main point of this function is not to set up `sources` if `sources` is not present we can assume the image does not have any available sources. Since this function is being called when the `restore` image is executed. At that point, we are not creating sources but restoring them so is safe to assume here if a `sources` is not present there's nothing to be restored (at least nothing custom created by this plugin) since this plugin is the one creating `sources` property.

felixarntz commented on modules/images/webp-uploads/image-edit.php:
```diff
@@ -320,10 +320,39 @@ function webp_uploads_get_next_full_size_key_from_backup( $attachment_id ) {
+		$backup_sources = array();
+	}
+
+	if ( ! defined( 'IMAGE_EDIT_OVERWRITE' ) || ! IMAGE_EDIT_OVERWRITE ) {
```
Let's not include support for this constant in this PR. That creates inconsistency because we don't support it elsewhere so far. There is an explicit TODO comment in the codebase to add support for it, and when we do that we should add support for the constant holistically. Right now you fix it here, but not when backing up sources, and it's better for now if we just keep this out.

In other words, let's always backup the sources regardless of the constant. We can later implement support for the constant separately.

felixarntz commented on modules/images/webp-uploads/image-edit.php:
```diff
@@ -320,10 +320,39 @@ function webp_uploads_get_next_full_size_key_from_backup( $attachment_id ) {
+		if ( empty( $backup_sources[ "full-{$target}" ] ) ) {
+			$backup_sources[ "full-{$target}" ] = $data['sources'];
+			update_post_meta( $attachment_id, '_wp_attachment_backup_sources', $backup_sources );
+		}
```
I think the implementation here is a bit unnecessarily complicated: Backing up the previously set sources in here can work exactly the same as when handling an image edit, i.e. we can reuse a lot of the existing logic we already have.

As far as I can tell, we should be able to re-use our `webp_uploads_backup_sources()` function just as is. Basically, we need to call that from `webp_uploads_update_attachment_metadata()` right before calling `webp_uploads_restore_image()`. And then we don't need to modify `webp_uploads_restore_image()` at all.

felixarntz reviewed (changes requested):
@mitogh This is a bit of a detail, but the implementation here has still some disparities for how WordPress core works. Try editing and restoring an image and compare the data in `_wp_attachment_backup_sizes` with the data in `_wp_attachment_backup_sources`.

Note how the hash used on the array keys in "sizes" is different from the hash used on the files themselves. However, in our own "sources" the hash on the entries is the same as the hash used on the files. So while that may feel correct, it is not the same as core's behavior.

So the custom implementation here is a bit off still. After a closer inspection, I think we can actually simplify this a lot _and_ fix it at the same time. Basically, we need to call our existing `webp_uploads_backup_sources()` function even when restoring an image. I've tested this locally, and with that change the result is as expected, with the hashes in `webp_uploads_backup_sizes` and `webp_uploads_backup_sources` matching correctly.

So this can be simplified quite a bit:
* The changes here in `webp_uploads_restore_image()` are not necessary.
* Instead, we need to add `$data = webp_uploads_backup_sources( $attachment_id, $data );` to the `case 'wp_restore_image'` condition in `webp_uploads_update_attachment_metadata()`, right before calling `webp_uploads_restore_image( $attachment_id, $data )`.
* The tests here need to be updated accordingly. We need to validate the hashes in `webp_uploads_backup_sizes` and `webp_uploads_backup_sources` match.

eugene-manuilov commented on modules/images/webp-uploads/image-edit.php:
```diff
@@ -320,10 +320,39 @@ function webp_uploads_get_next_full_size_key_from_backup( $attachment_id ) {
+				continue;
+			}
+
+			preg_match( '/-e(\d{13})/', $properties['file'], $matches );
+			if ( empty( $matches ) || count( $matches ) < 2 || empty( $matches[1] ) ) {
+				continue;
+			}
```
It's better to check the result of the `preg_match` function call rather than `$matches` array:
```suggestion
			$matches = array();
			if ( ! preg_match( '/-e(\d{13})/', $properties['file'], $matches ) ) {
				continue;
			}
```

mitogh replied:
Good call.

mitogh committed:
Remove non required conditionals

Just use the result from `preg_match` as the short circuit step to move on.

Co-authored-by: Eugene Manuilov <manuilov@google.com>

mitogh committed:
Merge branch 'trunk' into fix/295-backup-full-image-sources

mitogh committed:
Merge branch 'fix/295-backup-full-image-sources' of github.com:WordPress/performance into fix/295-backup-full-image-sources

mitogh committed:
Merge branch 'trunk' into fix/295-backup-full-image-sources

mxbclang commented:
@mitogh Just checking in on this!

mitogh commented:
Thanks for the ping @bethanylang  will wrap the work on this one over the weekend so is ready for a review on Monday :)

mitogh committed:
Merge branch 'trunk' into fix/295-backup-full-image-sources

mitogh committed:
Rename variable to match core naming.

Rename `$target` variable to `$suffix` in order to match
the same naming used in core.

mitogh committed:
Remove support for `IMAGE_EDIT_OVERWRITE` constant.

Don't add support for the constant that exists in core and
has not been fully supported here, all changes and behavior
for `IMAGE_EDIT_OVERWRITE` would be addressed later.

mitogh committed:
Remove duplicated code based on feedback.

Update the code by removing duplicated sections of code and
reuising existing sections that helps to do the same
logic.

Add new test case to review that the hashes matches existing
values from sizes.

Removal of non required tests due the constant is not currently
supported.

mitogh commented:
Just a heads up @bethanylang that this PR is ready for another review due the feedback from @felixarntz  has been addressed. 

Let me know in case you need additional feedback about this one, or if there's anything else I can help with on this one. 

Thanks.

felixarntz commented on modules/images/webp-uploads/image-edit.php:
```diff
@@ -387,9 +388,8 @@ function webp_uploads_get_next_full_size_key_from_backup( $attachment_id ) {
-
 	if ( ! is_array( $backup_sources ) ) {
-		return $data;
+		$backup_sources = array();
```
@mitogh Just double-checking, what was the particular reason you've added this change? I'm not saying it's wrong, but I would think this would also work without this change? Just wanna make sure I'm not missing something.

mitogh replied:
Sure, the main reason was that in case the backup source was not set up this would initialize it to a new empty array instead, in order to set up the references to the full-size images. 

Is an edge case, but wanted to make sure if the edge case happens the stored value is consistent, instead of having no value even if we have backup values available.

felixarntz reviewed (approved):
Thanks @mitogh, looks great! One follow-up question for you, would be great if you could have a quick look so that we can merge this in time for 1.2.0.

mukeshpanchal27 commented on modules/images/webp-uploads/image-edit.php:
```diff
@@ -387,9 +388,8 @@ function webp_uploads_get_next_full_size_key_from_backup( $attachment_id ) {
-
 	if ( ! is_array( $backup_sources ) ) {
-		return $data;
+		$backup_sources = array();
```
I have the same question as Felix and would like to explore that edge case.

I tried this edge case in which the `_wp_attachment_backup_sources` meta value is null or not an array. As per the below code snippet, it will initialize an empty array for `$backup_sources`.

	if ( ! is_array( $backup_sources ) ) {
		$backup_sources = array();
	}

After this snippet, we have another snippet that checks two things if `full-orig` key is not set in array `$backup_sources` or it's not an array.

	if ( ! isset( $backup_sources['full-orig'] ) || ! is_array( $backup_sources['full-orig'] ) ) {
		return $data;
	}

In this edge case the function return `$data` so as I understood it is worth initializing an empty array instead of returning the `$data` as pervious it does.

Please let me know if I'm getting anything wrong.

felixarntz replied:
From what I can tell it doesn't really matter though which way to go about this line of code, for the following reason:
* If `$backup_sources` is not an array, the following condition `! isset( $backup_sources['full-orig'] ) || ! is_array( $backup_sources['full-orig'] )` will also fore sure evaluate to `false`.
* So we can either initialize `$backup_sources` as empty array, causing `$data` to be returned in the following clause, or we can return `$data` right away. I don't think it changes the function behavior at all honestly.

cc @mitogh

mukeshpanchal27 reviewed (approved):
PR looks good to me and is ready to marge.
